rewrite-java/src/main/java/org/openrewrite/java/RenameVariable.java

```diff
@@ -38,35 +44,107 @@ public RenameVariable(J.VariableDeclarations.NamedVariable variable, String toNa
 
     private class RenameVariableByCursor extends JavaIsoVisitor<P> {
         private final Cursor scope;
-        private boolean isConstant = false;
+        private final Stack<Tree> currentNameScope = new Stack<>();
 
         public RenameVariableByCursor(Cursor scope) {
             this.scope = scope;
         }
 
         @Override
-        public J.CompilationUnit visitCompilationUnit(J.CompilationUnit cu, P p) {
-            return super.visitCompilationUnit(cu, p);
+        public J.Block visitBlock(J.Block block, P p) {
+            // * A variable name scope is owned by the ClassDeclaration regardless of what order it is declared.
+            // * Variables declared in a child block are owned by the corresponding block until the block is exited.
+            if (getCursor().getParent() != null && getCursor().getParent().getValue() instanceof J.ClassDeclaration) {
+                boolean isClassScope = block.getStatements().stream()
+                        .filter(it -> it instanceof J.VariableDeclarations)
+                        .anyMatch(it -> ((J.VariableDeclarations)it).getVariables().stream().anyMatch(v -> v.equals(variable)));
+
+                if (isClassScope && currentNameScope.isEmpty()) {
+                    // Defer the postVisit to pop the currentNameScope off the stack until the variable is found.
+                    currentNameScope.add(block);
+                }
+            }
+            return super.visitBlock(block, p);
         }
 
         @Override
         public J.Identifier visitIdentifier(J.Identifier ident, P p) {
-            if (!isConstant && ident.getSimpleName().equals(variable.getSimpleName()) &&
-                    isInSameNameScope(scope, getCursor()) &&
-                    !(getCursor().dropParentUntil(J.class::isInstance).getValue() instanceof J.FieldAccess)) {
+            // The size of the stack will be 1 if the identifier is in the right scope.
+            if (ident.getSimpleName().equals(variable.getSimpleName()) && isInSameNameScope(scope, getCursor()) && currentNameScope.size() == 1
+                    && !(getCursor().dropParentUntil(J.class::isInstance).getValue() instanceof J.FieldAccess)) {
```

Comment:
Binary operator on next line. Might want to run intellij formatter over whole file?